python/geoarrow/dataset.py

```diff
@@ -0,0 +1,139 @@
+    kernel = Kernel.box_agg(reader.schema.types[0])
+    for batch in reader:
+        kernel.push(batch.column(0))
+    return kernel.finish()
+
+
+def _index_fragments(fragments, columns, num_threads=None):
+    columns = list(columns)
+    if num_threads is None:
+        num_threads = _pa.cpu_count()
+
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        futures = []
+        for column in columns:
+            for fragment in fragments:
+                future = executor.submit(_index_fragment, fragment, column)
+                futures.append(future)
+
+        wait(futures)
+
+        num_fragments = len(futures) // len(columns)
+        results = []
+        for i, column in enumerate(columns):
+            results.append(
+                [futures[i * num_fragments + j].result() for j in range(num_fragments)]
+            )
+
+        metadata = [_pa.array(range(num_fragments))]
+        result_arrays = [_pa.concat_arrays(result) for result in results]
+        return _pa.table(metadata + result_arrays, names=["_fragment_index"] + columns)
+
+
+def _index_box_intersects(index, box, columns):
+    xmin, xmax, ymin, ymax = box.as_py().values()
+    expressions = []
+    for col in columns:
+        expr = (
+            (_compute.field(col, "xmin") <= xmax)
+            & (_compute.field(col, "xmax") >= xmin)
+            & (_compute.field(col, "ymin") <= ymax)
+            & (_compute.field(col, "ymax") >= ymin)
+        )
+        expressions.append(expr)
+
+    expr = expressions[0]
+    for i in range(1, len(expressions)):
+        expr = expr | expressions[i]
+
+    result = _ds.dataset(index).filter(expr).to_table()
+    return result.column(0)
+
+
+class GeoDataset:
+    def __init__(self, parent, *args, geometry_columns=None, **kwargs) -> None:
+        self._parent = None
+        self._index = None
+        self._geometry_columns = geometry_columns
+        self._fragments = None
+
+        if isinstance(parent, _ds.Dataset):
+            self._parent = parent
+        else:
+            self._parent = _ds.dataset(parent, *args, **kwargs)
+
+    def use_row_groups(self):
+        if isinstance(self._parent, _ds.FileSystemDataset) and isinstance(
+            self._parent.format, _ds.ParquetFileFormat
+        ):
+            row_group_fragments = []
+            for file_fragment in self._parent.get_fragments():
+                for row_group_fragment in file_fragment.split_by_row_group():
+                    row_group_fragments.append(row_group_fragment)
+
+            return GeoDataset(
+                _ds.FileSystemDataset(
+                    row_group_fragments, self._parent.schema, self._parent.format
+                )
+            )
+
+    @property
+    def schema(self):
+        return self._parent.schema
+
+    def get_fragments(self):
+        if self._fragments is None:
+            self._fragments = tuple(self._parent.get_fragments())
+
+        return self._fragments
```

Comment:
You can use [cached properties](https://docs.python.org/3/library/functools.html#functools.cached_property) if you don't want to check for initialization every time

Reply:
Ooh good question. Technically versions `<3.8` are [unsupported](https://devguide.python.org/versions/) now so IMO its okay to introduce 3.8 features. If we expect users to be on an older version of python/pyarrow regardless, then maybe better not to introduce these changes. Does the geospatial ecosystem stay relatively up to date?